### PR TITLE
5/n: Extract reference model call to plugins/accelerators

### LINKED ISSUE #4753: [FairScale] refactor plugin API to support FairScale Pipe Plugin

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Closes #4753. Final API change to allow custom DDP wrapper logic when extracting reference model.

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In in short, see following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by justusschock — APPROVED** [review]

**Comment by justusschock** [comment]
@SeanNaren I like it. This seems like a trivial function now, but could you maybe add a test for it to ensure that it still works, if we may come to more complicated cases?

**Comment by SeanNaren** [comment]
> @SeanNaren I like it. This seems like a trivial function now, but could you maybe add a test for it to ensure that it still works, if we may come to more complicated cases?

Absolutely, will do. Also ddp plugin tests are failing, need to ensure this logic is overridden correctly in the tests.

File: pytorch_lightning/trainer/connectors/model_connector.py [hunk 1]
```diff
@@ -55,6 +46,11 @@ def copy_trainer_model_properties(self, model):
...
-        model = self.trainer.model.module if is_dp_module else self.trainer.model
-        return model
+        return self._reference_model(self.trainer.model)
+
+    def _reference_model(self, model):
+        if self.trainer.accelerator_backend:
+            ref_model = self.trainer.accelerator_backend.reference_model(model)
+        else:
+            ref_model = model
+        return ref_model
```
Comment (SeanNaren) [inline pytorch_lightning/trainer/connectors/model_connector.py:56]: I'm not the biggest fan of having to deal with the model not being wrapped within the accelerator. I'm thinking of changing this to:

```python
if isinstance(model, LightningModule):
    return model
elif self.trainer.accelerator_backend:
    ref_model = self.trainer.accelerator_backend.reference_model(model)
else:
    raise RuntimeError("Model reference could not be found")
```

But I'm not sure what the edge cases are for this function.

Comment (SeanNaren) [reply pytorch_lightning/trainer/connectors/model_connector.py:56]: Going back and forth here but I think on par I'd prefer allowing the accelerator to deal with all logic here, this allows for more extensibility in the future within the accelerator and less prone to errors.

File: tests/trainer/properties/test_get_model.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    limit_train_batches = 2
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        limit_train_batches=limit_train_batches,
+        limit_val_batches=2,
+        max_epochs=1,
+        gpus=1,
+        accelerator='ddp_spawn'
+    )
+    trainer.fit(model)
```
Comment (tchaton) [inline tests/trainer/properties/test_get_model.py:105]: Can you add a ddp test with DDPLauncher ?

**Review by tchaton — CHANGES_REQUESTED** [review]
Can you add a DDP test ?

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 3]
```diff
@@ -108,3 +108,26 @@ def on_before_forward(self, model, *args):
...
+        Args:
+            model: Model with parallel wrapper.
+
+        Returns: Reference :class:`LightningModule` within parallel wrapper.
+
+        """
+        if isinstance(model, LightningDistributedDataParallel):
+            return model.module
+        else:
+            return model
```
Comment (Borda) [inline pytorch_lightning/plugins/ddp_plugin.py:133]:
```suggestion
        return model
```

File: pytorch_lightning/accelerators/dp_accelerator.py [hunk 4]
```diff
@@ -172,3 +174,9 @@ def reinit_scheduler_properties(self, optimizers: list, schedulers: list):
                 scheduler.__class__.__mro__[idx].__init__(scheduler, optimizer)
                 if state is not None:
                     scheduler.load_state_dict(state)
+
+    def reference_model(self, model: Union[LightningDataParallel, LightningModule]) -> LightningModule:
+        if isinstance(model, LightningDataParallel):
+            return model.module
+        else:
+            return model
```
Comment (Borda) [inline pytorch_lightning/accelerators/dp_accelerator.py:182]:
```suggestion
        return model
```

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 5]
```diff
@@ -108,3 +108,26 @@ def on_before_forward(self, model, *args):
 
     def optimizer_state(self, optimizer: Optimizer) -> dict:
         return optimizer.state_dict()
+
+    def reference_model_in_plugin_wrapper(
```
Comment (Borda) [inline pytorch_lightning/plugins/ddp_plugin.py:112]:
```suggestion
    def cast_model_from_plugin(
```
may we add something about what is the function doing in its name?

File: pytorch_lightning/accelerators/dp_accelerator.py [hunk 6]
```diff
@@ -172,3 +174,9 @@ def reinit_scheduler_properties(self, optimizers: list, schedulers: list):
                 scheduler.__class__.__mro__[idx].__init__(scheduler, optimizer)
                 if state is not None:
                     scheduler.load_state_dict(state)
+
+    def reference_model(self, model: Union[LightningDataParallel, LightningModule]) -> LightningModule:
```
Comment (Borda) [inline pytorch_lightning/accelerators/dp_accelerator.py:178]: shall it be rather a property or static?
since `self.trainer.model` is defined, this makes me think why it class method if `self` is not needed

**Review by Borda — COMMENTED** [review]
I have just concern about the methods which really do not require `self` if they shall be rather a class method or static

File: pytorch_lightning/accelerators/dp_accelerator.py (hunk 6 above)
Comment (SeanNaren) [reply pytorch_lightning/accelerators/dp_accelerator.py]: I wasn't able to make this a property, as the model isn't accessible from `self.trainer.model` in setting up ddp_train.

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 7]
```diff
@@ -108,3 +108,25 @@ def on_before_forward(self, model, *args):
 
     def optimizer_state(self, optimizer: Optimizer) -> dict:
         return optimizer.state_dict()
+
+    def module_from_plugin(
```
Comment (tchaton) [inline pytorch_lightning/plugins/ddp_plugin.py:112]: I find the name confusing here: module_from_plugin

**Review by tchaton — APPROVED** [review]

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 7 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/ddp_plugin.py:112]: would `model_from_plugin` be better? The issue is it's usually a module stored within the wrapper class, so ties to all parallel implementations already

File: pytorch_lightning/accelerators/ddp2_accelerator.py [hunk 8]
```diff
@@ -218,3 +218,6 @@ def sync_tensor(self,
                     group: Optional[Any] = None,
                     reduce_op: Optional[Union[ReduceOp, str]] = None) -> torch.Tensor:
         return sync_ddp_if_available(tensor, group, reduce_op)
+
+    def reference_model(self, model) -> LightningModule:
```
Comment (Borda) [inline pytorch_lightning/accelerators/ddp2_accelerator.py:222]: I guess we talke about...
```suggestion
    def get_reference_model(self, model) -> LightningModule:
```

File: pytorch_lightning/accelerators/dp_accelerator.py [hunk 9]
```diff
@@ -172,3 +174,10 @@ def reinit_scheduler_properties(self, optimizers: list, schedulers: list):
                 scheduler.__class__.__mro__[idx].__init__(scheduler, optimizer)
                 if state is not None:
                     scheduler.load_state_dict(state)
+
+    @property
+    def reference_model(self) -> LightningModule:
```
Comment (Borda) [inline pytorch_lightning/accelerators/dp_accelerator.py:179]: this is confusing as here it is the property and above is method

File: pytorch_lightning/trainer/connectors/model_connector.py [hunk 10]
```diff
@@ -55,6 +46,9 @@ def copy_trainer_model_properties(self, model):
             m.local_rank = self.trainer.local_rank
 
     def get_model(self):
-        is_dp_module = isinstance(self.trainer.model, (LightningDistributedDataParallel, LightningDataParallel))
-        model = self.trainer.model.module if is_dp_module else self.trainer.model
+        return self._reference_model(self.trainer.model)
+
+    def _reference_model(self, model):
```
Comment (Borda) [inline pytorch_lightning/trainer/connectors/model_connector.py:51]: I would add some verb here...

File: tests/trainer/properties/test_get_model.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import sys
+
+import pytest
+import torch
+
+from pytorch_lightning import Trainer
+from tests.backends.launcher import DDPLauncher
```
Comment (Borda) [inline tests/trainer/properties/test_get_model.py:20]: cc: @tchaton

File: pytorch_lightning/accelerators/dp_accelerator.py (hunk 9 above)
Comment (SeanNaren) [reply pytorch_lightning/accelerators/dp_accelerator.py:179]: this is a mistake, let me fix

**Review by Borda — APPROVED** [review]
